### PR TITLE
Fix txn commit hooks executed on dry run

### PR DESCRIPTION
**Pull request by pratikgajjar**
**Describe the bug**
TLDR; When import export is used it doesn’t remove commit hooks on dry_run. 
We use Django 2.2 with Postgres (latest). We use transaction commit hooks to add celery jobs. When It is used post_save signal doesn’t add background job twice. 

https://docs.djangoproject.com/en/3.1/topics/db/transactions/#django.db.transaction.on_commit
https://docs.djangoproject.com/en/3.1/topics/db/transactions/#savepoints
https://github.com/django-import-export/django-import-export/issues/1078

**To Reproduce**
1.  Register post_save signal for a model
```py
@receiver([post_save, post_delete], sender=ContactCohortMap)
@disable_for_loaddata
def contact_cohort_map_signal(sender, instance, **kwargs):
    if instance:
        transaction.on_commit(lambda: some_celery_task.delay('arg1'))
```
2. Add background job in celery or any function using commit hooks
```py
transaction.on_commit(lambda: some_celery_task.delay('arg1'))
```
3. Import some valid data
4. Background job will be added in celery twice - one when it dry runs other is successful commit

**Versions**
 - Django Import Export: 1.2.0
 - Python 3.7
 - Django 2.2.5

**Expected behavior**
Execute commit hooks on successful db commit.

**Screenshots**
Can show if requested.

**Additional context**
When import export used with transaction support. It creates a savepoint then execute each row in txn.atomic() block. 
```py
    def import_data_inner(self, dataset, dry_run, raise_errors, using_transactions, collect_failed_rows, **kwargs):
        result = self.get_result_class()()
        result.diff_headers = self.get_diff_headers()
        result.total_rows = len(dataset)

        if using_transactions:
            # when transactions are used we want to create/update/delete object
            # as transaction will be rolled back if dry_run is set
            sp1 = savepoint()
       . . .
        for i, row in enumerate(dataset.dict, 1):
            with atomic_if_using_transaction(using_transactions):
                row_result = self.import_row(
                    row,
                    instance_loader,
                    using_transactions=using_transactions,
                    dry_run=dry_run,
                    **kwargs
                )
      . . .
        if using_transactions:
            if dry_run or result.has_errors():
                savepoint_rollback(sp1)
            else:
                savepoint_commit(sp1)
```
When you execute db query in nested atomic blocks it adds commit hook function in child or intermediate savepoints.
```py
    def savepoint_rollback(self, sid):
        """
        Roll back to a savepoint. Do nothing if savepoints are not supported.
        """
        if not self._savepoint_allowed():
            return

        self.validate_thread_sharing()
        self._savepoint_rollback(sid)

        # Remove any callbacks registered while this savepoint was active.
        self.run_on_commit = [
            (sids, func) for (sids, func) in self.run_on_commit if sid not in sids
        ]
```

So when dry run was done it didn’t remove callback for intermediate save points

**Fix without updating library** 
```py
class BaseModelResource(resources.ModelResource):
    save_points = []

    def before_save_instance(self, instance, using_transactions, dry_run):
        """
        Override to add additional logic. Does nothing by default.
        Saving intermediate savepoints of txn
        """
        if using_transactions and dry_run:
            con = transaction.get_connection()
            self.save_points.extend(con.savepoint_ids)

    def after_import(self, dataset, result, using_transactions, dry_run, **kwargs):
        """
        Override to add additional logic. Does nothing by default.
        Manually removing commit hooks for intermediate savepoints of atomic transaction
        """
        if using_transactions and dry_run:
            con = transaction.get_connection()
            for sid in self.save_points:
                con.run_on_commit = [
                    (sids, func) for (sids, func) in con.run_on_commit if sid not in sids
                ]
        super().after_import(dataset, result, using_transactions, dry_run, **kwargs)
```
Inherit this class to avoid this problem.

 
**Solution**
Saved all intermediate savepoints and removed commit hooks from them

**Acceptance Criteria**
Will add later

Edit : 
Closed PR : https://github.com/django-import-export/django-import-export/pull/1176
Because it was created from master branch

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -60,6 +60,6 @@ install:
   - pip install coveralls
 script:
    - PYTHONPATH=".:tests:$PYTHONPATH" python -Wall -m coverage run --omit='setup.py' --source=. tests/manage.py test core --settings=
-   - if python -c 'import sys; sys.exit(1 - (sys.version_info >= (3, 6)))'; then isort --check-only import_export tests; fi
+   - if python -c 'import sys; sys.exit(1 - (sys.version_info >= (3, 6)))'; then isort --check-only -rc import_export tests; fi
```
Comment (andrewgy8) [inline .travis.yml:63]: Can we remove the `rc` from here? Its adding a bit of noise to the PR. If we want this in isort, we should add it in another PR.

**Comment by adamchainz** [comment]
I don't think this is a good approach. Fiddling with `run_on_commit` is not documented behaviour inside Django. The only reliable way to prevent these `on_commit` hooks from firing is to *always* use an `@atomic` and roll it back. As I understand `@atomic` is only used if the user sets the setting/class attribute to true, which is imo a bad default.

**Comment by pratikgajjar** [comment]
@adamchainz Thanks for your feedback. Yes, you are right this all happening because library uses manual save point tracking to do rollback on dry run when using transactions. So instead of using [savepoint_rollback](https://docs.djangoproject.com/en/3.1/topics/db/transactions/#savepoint-rollback) we should use [rollback](https://docs.djangoproject.com/en/3.1/topics/db/transactions/#transaction-rollback). This is what you are suggesting right ? 

```py
    # file : import_export/resources.py 
    # line no : 588
    def import_data_inner(self, dataset, dry_run, raise_errors, using_transactions, collect_failed_rows, **kwargs):
        . . .
        if using_transactions:
            # when transactions are used we want to create/update/delete object
            # as transaction will be rolled back if dry_run is set
            sp1 = savepoint()
 
 ```

**Comment by adamchainz** [comment]
I think if you use `atomic()`, you should be able to use [`set_rollback`](https://docs.djangoproject.com/en/3.0/topics/db/transactions/#django.db.transaction.set_rollback) to tell Django to roll back, and then when you exit the atomic it will do so. Afraid I can't give much more guidance on this right now, so my comments are a bit driveby.

**Comment by andrewgy8** [comment]
Thanks for the contrib. Im going to convert this to a draft since it is lacking test coverage and documentation.

**Comment by cachitas** [comment]
@pratikgajjar 
Your proposed solution of inheriting from your `BaseModelResource` is not working for me. My signal still runs twice.
Is it required to have the receiver using ` transaction.on_commit` as you show on your case?
I am not using celery tasks.

Anyone else tried to test this solution and got it working?

**Comment by pratikgajjar** [comment]
> @pratikgajjar
> Your proposed solution of inheriting from your `BaseModelResource` is not working for me. My signal still runs twice.
> Is it required to have the receiver using ` transaction.on_commit` as you show on your case?
> I am not using celery tasks.
> 
> Anyone else tried to test this solution and got it working?

@cachitas  
Yes you need to use transaction.on_commit over your function, Celery task was just an example.

```py
transaction.on_commit(lambda: your_function(args))
```

**Comment by cachitas** [comment]
Thanks @pratikgajjar, that did the trick.
I have been playing around with the transaction but I am not comfortable enough with it. Also, from @adamchainz comment it looks like we may need to rethink the current implementation.

For now I decided to go with @andrewgy8 suggestion of disconnecting the signal temporarily. It does not feel as robust, but it fits my needs. Here is a possible solution in case someone needs inspiration.

```python
# resources.py
from django.db.models.signals import post_save
from .signals import some_callback

class BaseModelResource(ModelResource):
    """Inherit your Resources from this Base class."""
    
    to_reconnect = []  # [(signal, sender, handler), ...]

    def before_save_instance(self, instance, using_transactions, dry_run):
        # adapt this to your needs
        signal = post_save
        sender = instance._meta.model
        handler = some_callback

        if dry_run:
            disconnected = signal.disconnect(receiver=handler, sender=sender)

            if disconnected:
                self.to_reconnect.append((signal, sender, handler))

    def after_save_instance(self, instance, using_transactions, dry_run):
        # no need to check if we are doing a dry_run here
        # only the previously disconnected signals will be reconnected here
        for signal, sender, handler in self.to_reconnect:
            signal.connect(handler, sender)
```

Funny Fact: while working on this, I found a bug in `Signal.disconnect()` [PR](https://github.com/django/django/pull/14187)

**Comment by pratikgajjar** [comment]
> Thanks @pratikgajjar, that did the trick.
> I have been playing around with the transaction but I am not comfortable enough with it. Also, from @adamchainz comment it looks like we may need to rethink the current implementation.
> 
> For now I decided to go with @andrewgy8 suggestion of disconnecting the signal temporarily. It does not feel as robust, but it fits my needs.
> 
> Funny Fact: while working on this, I found a bug in `Signal.disconnect()` [PR](https://github.com/django/django/pull/14187)

When you disconnect a signal does it happen for only that execution or Like in parallel some other (thread) code is modifying the model then it also gets disconnected.

You can test that with below flow
1. Create one api endpoint which modifies your model
2. Add sleep after disconnecting signal 
3. From django admin modify this model.
4. Hit api endpoint  (3. Takes row level lock so i don't think you will face any db lock related issue)

Ideally it should trigger signal. 

Test it with python manage.py runserver (single threaded so it will block 4th step) and with gunicorn or uwsgi.

**Comment by pratikgajjar** [comment]
Another way to solve this problem is not to use multiple savepoints for dry run with db txn. It requires good understanding of this library. Not sure what else it will break.  

I have gone through step by step each line of code django executes to fix this problem. The method I have used is also used by django internally to rollback a txn. 

We are using this fix in production for good amount of time and we haven't faced any issues.

**Comment by cachitas** [comment]
> When you disconnect a signal does it happen for only that execution or Like in parallel some other (thread) code is modifying the model then it also gets disconnected.

Indeed with a `sleep()` after disconnecting, modifying the model from other sources did not trigger the signal using `runserver`.
Using `gunicorn` with 4 workers the signal triggered correctly during `sleep()`.

**Comment by pratikgajjar** [comment]
> > When you disconnect a signal does it happen for only that execution or Like in parallel some other (thread) code is modifying the model then it also gets disconnected.
> 
> Indeed with a `sleep()` after disconnecting, modifying the model from other sources did not trigger the signal using `runserver`.
> Using `gunicorn` with 4 workers the signal triggered correctly during `sleep()`.

So exectuion depends on how you run your code. It means within worker process it will disable the signal. If same worker had served that requests then it also wouldn't have triggered signal.

Thank you for testing it.

**Comment by matthewhegarty** [comment]
Hi Pratik - this has been around for a while.  Do you want us to resurrect it for consideration for inclusion in the library?  If so, we'll have to have a discussion about whether it is still required.  We should do this before embarking on any major work.

**Comment by matthewhegarty** [comment]
Closing due to inactivity
